src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Diagnostics/StructuredAnalyzerConfigOptions.cs

```diff
@@ -0,0 +1,102 @@
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using Microsoft.CodeAnalysis.Diagnostics.Analyzers.NamingStyles;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Diagnostics;
+
+/// <summary>
+/// <see cref="AnalyzerConfigOptions"/> that memoize structured (parsed) form of certain complex options to avoid parsing them multiple times.
+/// Storages of these complex options may directly call the specialized getters to reuse the cached values.
+/// </summary>
+internal abstract class StructuredAnalyzerConfigOptions : AnalyzerConfigOptions
+{
+    internal sealed class Implementation : StructuredAnalyzerConfigOptions
+    {
+        private readonly AnalyzerConfigOptions _options;
+        private readonly Lazy<NamingStylePreferences> _lazyNamingStylePreferences;
+
+        public Implementation(AnalyzerConfigOptions options)
+        {
+            _options = options;
+            _lazyNamingStylePreferences = new Lazy<NamingStylePreferences>(() => EditorConfigNamingStyleParser.ParseDictionary(_options));
+        }
+
+        public override bool TryGetValue(string key, [NotNullWhen(true)] out string? value)
+            => _options.TryGetValue(key, out value);
+
+        public override IEnumerable<string> Keys
+            => _options.Keys;
+
+        public override NamingStylePreferences GetNamingStylePreferences()
+            => _lazyNamingStylePreferences.Value;
+    }
+
+    public abstract NamingStylePreferences GetNamingStylePreferences();
+
+    public static StructuredAnalyzerConfigOptions Create(ImmutableDictionary<string, string> options)
+    {
+        Contract.ThrowIfFalse(options.KeyComparer == KeyComparer);
+        return new Implementation(new DictionaryAnalyzerConfigOptions(options));
+    }
+
+    public static bool TryGetStructuredOptions(AnalyzerConfigOptions configOptions, [NotNullWhen(true)] out StructuredAnalyzerConfigOptions? options)
+    {
+        if (configOptions is StructuredAnalyzerConfigOptions structuredOptions)
+        {
+            options = structuredOptions;
+            return true;
+        }
+
+#if CODE_STYLE
+        if (TryGetCorrespondingCodeStyleInstance(configOptions, out options))
+        {
+            return true;
+        }
+#endif
+
+        options = null;
+        return false;
+    }
+
+#if CODE_STYLE
+    // StructuredAnalyzerConfigOptions is defined in both Worksapce and Code Style layers. It is not public and thus can't be shared between these two.
+    // However, Code Style layer is compiled against the shared Workspace APIs. The ProjectState creates and holds onto an instance
+    // of Workspace layer's version of StructuredAnalyzerConfigOptions. This version of the type is not directly usable by Code Style code.
+    // We create a clone of this instance typed to the Code Style's version of StructuredAnalyzerConfigOptions.
```

Comment:
:( this codestyle stuff turns my brain into goo

Reply:
Yes. It's very complex. Hopefully after next few PRs it's gonna get a bit simpler.